### PR TITLE
DEV: Extra Pre-commit Checks

### PR DESCRIPTION
**Pull request by John-P**
I have compiled many extra checks for pre-commit, some of which are the same as those used by deepsource. Proposed new additions to pre-commit:

- isort: Sort and group imports (stdlibrary, 3rd party, current package)
- flake8-isort: Check import ordering
- flake8-comprehensions: Suggestions for better comprehensions
- flake8-mutable: Check for mutable default arguments
- flake8-use-fstring: Use of f-strings vs old style format strings
- flake8-simplify: Suggestions to simplify code
- pep8-naming: Check PEP8 class naming
- flake8-eradicate: Find dead/commented out code
- flake8-builtins: Check for built-ins being used as variables
- flake8-cognitive-complexity: Check max function complexity (set in setup.cfg)
- flake8-expression-complexity: Check max expression complexity (set in setup.cfg)
- flake8-return: Check return statements (no redundant assignments)
- flake8-spellcheck: Spelling checker (comments only, new words in whitelist.txt)
- flake8-pytest-style: Check against pytest style guide
- flake8-fixme: Check for FIXME, TODO, and XXX left in comments
- flake8-sql: Check SQL statement style
- nitpick: Check that black, flake8, isort have compaible options
- flake8-logging-format: Validate (lack of) logging format strings
- flake8-pie: Misc. linting rules

Of course we don't have to use all of these but I went through a bunch of them and tried them out. These were the ones that looked helpful and did not have any obvious annoying bugs or conflicts.

Some of these could be added to travis also if that would be useful.

### CODE REVIEW

File: setup.cfg [hunk 1]
```diff
@@ -18,10 +18,20 @@ universal = 1
...
 
 [tool:pytest]
 collect_ignore = ['setup.py']
 
+[isort]
+profile = black
+multi_line_output = 3
+skip_gitignore = True
+line_length = 88
+filter_files = True
```
Comment (shaneahmed) [inline setup.cfg:37]: empty line?

**Review by shaneahmed — COMMENTED** [review]
Thanks @John-P I am in favour of these pre-commits. My only question, can we skip some errors from these packages to allow commit in case the programmer would like to force skip?

File: setup.cfg (hunk 1 above)
Comment (John-P) [reply setup.cfg:37]: empty line??

**Comment by John-P** [comment]
> Thanks @John-P I am in favour of these pre-commits. My only question, can we skip some errors from these packages to allow commit in case the programmer would like to force skip?

So, Looking into the flake8 and pre-commit configs/options. There are a few things we could do:
1. Run a subset of checks on commit and more checks on push. You can customise which commands run on any of: commit, push, writing commit message, checkout, merge, etc.
2. You can use `$ commit --no-verify` to by pass the hooks for a single commit.
3. We could also split the checks into two groups, one of which will always return true (and set verbose true to still print, as passing tests don't print by default) and therefore not ever cause the action to block but will still print out a message.

Example Pre-Commit Hook That Will Always Print No Mater The Exit Code:

```yaml
  - repo: local
    hooks:
      - id: say-hello
        name: Say Hello
        description: Just prints a message.
        entry: python -c "print('Hello there!')"
        language: python
        verbose: true
```

While looking at the documentation, I also noticed that pre-commit can now auto set up when the repo is cloned. This is probably suited for a different PR though.

**Review by shaneahmed — APPROVED** [review]

**Pull request merged**